feat: Send PDF invoice as mail attachement

adrianmoloce opened the pull request:
- Test sendmail was called with invoice_name
  - Render invoice template  and attach it to the mail message
  - TO DO: Mock rendering as it slowed down tests and generates spam invoices
  - TO DO: Mock with open invoice file path if rendering is mocked

amoloce committed:
feat: Send PDF incoice as mail attachement

  - Test sendmail was called with invoice_name
  - Render invoice template  and attach it to the mail message
  - TO DO: Mock rendering as it slowed down tests and generates spam
    invoices
  - TO DO: Mock with open invoice file path if rendering is mocked

adrianmoloce commented on src/invoice_utils/web.py:
```diff
@@ -14,10 +18,12 @@
 
 from invoice_utils.engine import InvoicingEngine
 from invoice_utils.models import InvoicedItem
+load_dotenv() # Need to do this now for it to work.
```
Needed to do this here to be able to run the server manually, is there a better way to write this, not within the imports?
But even after the fix it seems like I can't send mails with gmail server anymore.

koosie0507 commented on src/invoice_utils/web.py:
```diff
@@ -83,20 +89,24 @@ def generate_invoice(request: InvoiceRequest):
     )
+    renderer = PdfInvoiceRenderer("invoice")
+    invoice_path = root_dir / f"{request.header.timestamp:%Y%m%d}-{int(request.header.number):04}-invoice.pdf"
+    renderer.render(context, str(invoice_path))
```
What happens to the file that was generated after the HTTP response was sent to the client? I would expect it's being removed. A `ContextManager` would be nice here.

koosie0507 replied:
Another topic relates to having write access on disk. Please test that not having write access on disk doesn't crash the web endpoint. The solution is completely up to you and can range from using streams instead of files to simply logging a warning related to not having write access.
The response differs depending on the chosen solution. If you want to keep using files, the server should return 503 or 507 depending on the reason why it can't write the file to disk.
Of course, the tests should be written first.

adrianmoloce replied:
For this endpoint, the invoice creation on disk is not required so I updated the render method with a flag 'persist'. I created the functionality to check the write access of a configurable location to persist on disk, which may be used in the future. Perhaps this endpoint is doing too much at the moment and in the future could be split into an invoice generation endpoint and a send_invoice\invoice_id endpoint.

koosie0507 commented on tests/test_web.py:
```diff
@@ -241,5 +232,31 @@ def test_email_body_was_sent_with_expected_body(
+    ],
+    indirect=["environment"]
+)
+def test_sendmail_was_called_with_pdf_attachment(environment, http, server, email_invoice_request_body):
```
Does the test generate a file on disk? If so, find a way that it doesn't. Unit tests need to be able to run in isolation, on platforms that don't have disk access.
I suggest you have a look at our good friends `patch` and `MagicMock`.

adrianmoloce replied:
I patched the render method to save time, although in the end I chose to not create any files at this point as it's not necessary for now.

koosie0507 commented on src/invoice_utils/web.py:
```diff
@@ -83,20 +89,24 @@ def generate_invoice(request: InvoiceRequest):
         int(request.header.number), request.header.timestamp, request.items
     )
+    renderer = PdfInvoiceRenderer("invoice")
+    invoice_path = root_dir / f"{request.header.timestamp:%Y%m%d}-{int(request.header.number):04}-invoice.pdf"
```
It often happens that the web app is installed in a directory it does not have write access to. When that happens, the app usually allows configuring a directory with read/write access where it can perform file manipulations.
Introduce an environment variable and use the directory specified by the environment variable when one is specified.

adrianmoloce replied:
I created a new env variable for this. Does the responsibility for the directory to exist and have write access then fall to the one using the application or the server? So far I've created the env var and a default dir. I'm still using the root_dir path with the new directory, should I drop that and allow the directory to be anywhere on the server?

amoloce committed:
Merge branch 'feat/send-mail-body' of github.com:quick-invoice-api/invoice-utils into feat/mail-pdf-attachment

amoloce committed:
feat: Configure which dir to use for invoice generation

  - Add env var for configuring the dir for invoice generation
  - Add test which checks the render method was called with expected
    path

amoloce committed:
feat: Respond with 507 if invoice dir does not exist

amoloce committed:
feat: Respond with 507 if invoice dir does not exist

amoloce committed:
Merge branch 'feat/mail-pdf-attachment' of github.com:quick-invoice-api/invoice-utils into feat/mail-pdf-attachment

amoloce committed:
feat: Respond with 507 if cant write to invoice dir

  - Return a 507 if the configured directory doesnt have write access

amoloce committed:
fix: Mock the invoice rendering

  - Patch the render method to just return bytes content
  - Remove extra tab in test definitions

amoloce committed:
feat: Add optional flag to persist the invoice to disk.

  - Added optional flag persist to store the invoice on disk
  - Render method now returns the content of the invoice
  - Added flag to previous usage of the render method

amoloce committed:
feat: Send invoice endpoint now uses streams to send invoice.

  - Using the new render method, the endpoint sends data without
    creating the invoice on disk
  - Validation methods exist for future usage which stores the files

koosie0507 commented on tests/test_web.py:
```diff
@@ -225,16 +227,109 @@ def test_smtp_starttls_env_flag_must_be_true_boolean(environment, http, server,
+    http.post("/invoice", json=email_invoice_request_body)
+    timestamp = datetime.fromisoformat(email_invoice_request_body['header']['timestamp'])
+    invoice_number = int(email_invoice_request_body['header']['number'])
+    invoice_name = f"{timestamp:%Y%m%d}-{invoice_number:04}-invoice.pdf"
```
replace this logic with a constant please.

koosie0507 commented on tests/test_web.py:
```diff
@@ -225,16 +227,109 @@ def test_smtp_starttls_env_flag_must_be_true_boolean(environment, http, server,
+    ],
+    indirect=["environment"]
+)
+def test_sendmail_was_called_with_pdf_attachment(
```
I see that we're testing whether sendmail is called. How about adding a test when sendmail fails with an `Exception`.

adrianmoloce replied:
Yes, the test at line 127 checks this, any error raised within the smtplib context is handeled with a specific message.

koosie0507 commented on tests/test_web.py:
```diff
@@ -225,16 +227,109 @@ def test_smtp_starttls_env_flag_must_be_true_boolean(environment, http, server,
+        with mocker.patch("invoice_utils.web.os.access", return_value=False):
+            res = http.post("/invoice", json=email_invoice_request_body)
+    assert res.status_code == 507
+    assert res.json() == {"detail": "Invoices directory does not have write access."}
```
I like the test and the design. The error message could be less descriptive in this case. Should be something like 'insufficient rights to store invoice'.

koosie0507 commented on tests/test_web.py:
```diff
@@ -225,16 +227,109 @@ def test_smtp_starttls_env_flag_must_be_true_boolean(environment, http, server,
+):
+    res = http.post("/invoice", json=email_invoice_request_body)
+    assert res.status_code == 507
+    assert res.json() == {"detail": "Invoices directory not set up."}
```
I like the test and the design. The message could be different. Something along the lines of 'no local storage available for invoices'.

koosie0507 reviewed (commented):
This already looks better than most code that sends e-mails :-D.

amoloce committed:
fix: Resolved code review comments

  - Rewrote 2 error messages for the invoice storage
  - Removed logic from test
